File: src/main/java/com/gmail/borlandlp/minigamesdtools/gui/hotbar/Hotbar.java
package com.gmail.borlandlp.minigamesdtools.gui.hotbar;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.items.SlotItem;
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.utils.Leveling;
import net.minecraft.server.v1_12_R1.PacketPlayOutExperience;
import org.bukkit.craftbukkit.v1_12_R1.entity.CraftPlayer;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;

public abstract class Hotbar {
    protected SlotItem[] slots = new SlotItem[9];
    protected Deque<SlotItem> itemsInQueue = new ArrayDeque<>();
    protected Player player;

    public Hotbar(Player player) {
        this.player = player;
    }

    public Hotbar(Player player, SlotItem[] items) {
        this.player = player;
        this.slots = items;
    }

    public void addSlot(SlotItem slotItem) {
        this.itemsInQueue.add(slotItem);
    }

    public void setSlot(int index, SlotItem slot) throws Exception {
        if(index < this.slots.length && index >= 0) {
            this.slots[index] = slot;
        } else {
            throw new Exception("HotbarSlot with ID: '" + slot.getName() + "' has an incorrect index[" + index + "] or is out of bounds [correct -> 0-8]");
        }
    }

    public void update() {
        this.updateSlotsQueue();
    }

    public void performAction(int slotID) {
        Debug.print(Debug.LEVEL.NOTICE,"[SkyBattle] performAction slotID:" + slotID + " for player:" + this.player.getDisplayName());
        if(slotID < this.slots.length && slotID >= 0 && this.slots[slotID] != null) {
            this.slots[slotID].performClick(this.player);
            if(this.slots[slotID].getAmount() < 1) {
                this.slots[slotID] = null;
            }
        }
    }

    public ItemStack[] getDrawData() {
        ItemStack[] data = new ItemStack[9];
        for(int i = 0; i < this.slots.length; i++) {
            if(this.slots[i] != null) {
                data[i] = this.slots[i].getDrawData();
            }
        }

        return data;
    }

    public void draw() {
        ItemStack[] drawData = this.getDrawData();
        ItemStack[] inventory = this.player.getInventory().getContents();

        int heldSlot = player.getInventory().getHeldItemSlot();
        if(this.slots[heldSlot] != null && this.slots[heldSlot].inCooldown()) {
            float percent = ((float)(this.slots[heldSlot].getCooldownTime() - this.slots[heldSlot].getCooldownRemain()) / this.slots[heldSlot].getCooldownTime());
            int secRemain = (int) (this.slots[heldSlot].getCooldownRemain() / 1000);
            Leveling.ExperienceContainer value = Leveling.calculateWithPercentage((int) (this.slots[heldSlot].getCooldownRemain() / 1000), percent);
            PacketPlayOutExperience packet = new PacketPlayOutExperience(percent, (int) value.total_exp, secRemain);

            ((CraftPlayer) player).getHandle().playerConnection.sendPacket(packet);
        }

        if(!isIdentDrawData(drawData, inventory)) {
            for(int i = 0; i < 9; i++) {
                inventory[i] = drawData[i];
            }
            player.getInventory().setContents(inventory);
        }
    }

    private boolean isIdentDrawData(ItemStack[] drawData, ItemStack[] inventory) {
        for(int i = 0; i < 9; i++) {
            if((inventory[i] == null && drawData[i] != null) || (inventory[i] != null && drawData[i] == null)) {
                return false;
            } else if(inventory[i] != null && drawData[i] != null) {
                boolean materialChanged = inventory[i].getType() != drawData[i].getType();
                boolean amountChanged = inventory[i].getAmount() != drawData[i].getAmount();
                if(materialChanged || amountChanged) {
                    return false;
                }
            }
        }

        return true;
    }

    private void updateSlotsQueue() {
        for (int i = 0; i < this.slots.length-1; i++) { // use 8 slots. 9 slot - reserved for correct work itemHeldEvent
            if(this.slots[i] == null && this.itemsInQueue.size() > 0) {
                this.slots[i] = this.itemsInQueue.poll();
            }
        }
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/gui/hotbar/Hotbar.kt
package com.gmail.borlandlp.minigamesdtools.gui.hotbar

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.items.SlotItem
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.utils.Leveling.calculateWithPercentage
import net.minecraft.server.v1_12_R1.PacketPlayOutExperience
import org.bukkit.craftbukkit.v1_12_R1.entity.CraftPlayer
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import java.util.*

abstract class Hotbar {
    protected var slots = arrayOfNulls<SlotItem>(9)
    protected var itemsInQueue: Deque<SlotItem> = ArrayDeque()
    protected var player: Player

    constructor(player: Player) {
        this.player = player
    }

    constructor(player: Player, items: Array<SlotItem?>) {
        this.player = player
        slots = items
    }

    fun addSlot(slotItem: SlotItem) {
        itemsInQueue.add(slotItem)
    }

    @Throws(Exception::class)
    fun setSlot(index: Int, slot: SlotItem) {
        if (index < slots.size && index >= 0) {
            slots[index] = slot
        } else {
            throw Exception("HotbarSlot with ID: '" + slot.name + "' has an incorrect index[" + index + "] or is out of bounds [correct -> 0-8]")
        }
    }

    open fun update() {
        updateSlotsQueue()
    }

    fun performAction(slotID: Int) {
        Debug.print(
            Debug.LEVEL.NOTICE,
            "[SkyBattle] performAction slotID:" + slotID + " for player:" + player.displayName
        )
        if (slotID < slots.size && slotID >= 0 && slots[slotID] != null) {
            slots[slotID]!!.performClick(player)
            if (slots[slotID]!!.amount < 1) {
                slots[slotID] = null
            }
        }
    }

    val drawData: Array<ItemStack?>
        get() {
            val data = arrayOfNulls<ItemStack>(9)
            for (i in slots.indices) {
                if (slots[i] != null) {
                    data[i] = slots[i]!!.drawData
                }
            }
            return data
        }

    fun draw() {
        val drawData = drawData
        val inventory = player.inventory.contents
        val heldSlot = player.inventory.heldItemSlot
        if (slots[heldSlot] != null && slots[heldSlot]!!.inCooldown()) {
            val percent =
                (slots[heldSlot]!!.getCooldownTime() - slots[heldSlot]!!.cooldownRemain).toFloat() / slots[heldSlot]!!.getCooldownTime()
            val secRemain = (slots[heldSlot]!!.cooldownRemain / 1000).toInt()
            val value =
                calculateWithPercentage((slots[heldSlot]!!.cooldownRemain / 1000).toInt(), percent)
            val packet = PacketPlayOutExperience(percent, value.total_exp.toInt(), secRemain)
            (player as CraftPlayer).handle.playerConnection.sendPacket(packet)
        }
        if (!isIdentDrawData(drawData, inventory)) {
            for (i in 0..8) {
                inventory[i] = drawData[i]
            }
            player.inventory.contents = inventory
        }
    }

    private fun isIdentDrawData(
        drawData: Array<ItemStack?>,
        inventory: Array<ItemStack?>
    ): Boolean {
        for (i in 0..8) {
            if (inventory[i] == null && drawData[i] != null || inventory[i] != null && drawData[i] == null) {
                return false
            } else if (inventory[i] != null && drawData[i] != null) {
                val materialChanged = inventory[i]!!.type != drawData[i]!!.type
                val amountChanged = inventory[i]!!.amount != drawData[i]!!.amount
                if (materialChanged || amountChanged) {
                    return false
                }
            }
        }
        return true
    }

    private fun updateSlotsQueue() {
        for (i in 0 until slots.size - 1) { // use 8 slots. 9 slot - reserved for correct work itemHeldEvent
            if (slots[i] == null && itemsInQueue.size > 0) {
                slots[i] = itemsInQueue.poll()
            }
        }
    }
}